_posts/2016-05-19-big-data-cap-theorem-plain-english.md

```diff
@@ -0,0 +1,111 @@
+
+As you have imagined, this is what it's called "scaling up" or "scaling vertically"[^scale]. You find your bottleneck (e.g. memory) and just make it bigger.
+
+Someday you'll see that the cost of expanding your library is prohibitive or, directly that it is physically impossible to add more shelves, indexes or doors. That day is the day you realise you need to build another library.
+
+And yes, your guess is correct, creating a library network and adding more nodes (buildings) to it would be to "scale out" or "scale horizontally" your library.
+
+## Sharding, redundancy and replication factor
+
+### Sharding
+Now that you have multiple buildings you need to take some important decisions. The first one is to decide how to distribute your books among the different libraries (sharding[^shard]).
+
+One possible and easy solution would be to distribute them by author. Books of authors whose surname starts with A, B, C or D go to building number 1. Authors with surname starting with E, F, G or H go to building number 2 and so on.
+
+You soon realise a problem with that approach. The authors surnames are not distributed equally and some authors (like Terry Pratchet) have written many more books than others. The implication is that some of your buildings are half empty while some others are too busy.
+
+You need to choose a better criterion (partition or shard key) that is easy to compute and well distributed. There are several options and there is not a good answer that will work in any scenario (it will depend on how your index is created, what queries can you do, what are the most common queries people is doing...). But, as an example, hashing the book title and using the modulus to decide which building the book is in would provide a better distribution of the books.
+
+### Redundancy and replication factor
+
+One day you read about [what happened to the alexandrians](https://en.wikipedia.org/wiki/Destruction_of_the_Library_of_Alexandria) and think about what would happen if one of the buildings catches fire. So you decide to buy extra copies of your books and put them in different buildings (redundancy).
+
+Of course, this comes with a cost and implications, both of them will depend on how many copies you want to keep (replication factor) among others.
+
+Besides the obvious advantage of having backup copies of each book (you could have achieved that by simply having a warehouse), this new architecture has a major drawback: every time you receive a new edition of a book, you'll need to send the updated copy to each of the buildings with a van, a truck, a train or a bicycle (unfortunately, books don't yet send themselves!)
+
+## CAP Theorem (aka Brewer's theorem)[^cap]
+
+Now that you have several buildings spread all over the city with the books distributed among them is when you'll start having troubles.
```

Comment:
Just trying to make sure this sentence reads correctly, as I think there's something a bit off about the "among them is" part:

> Once you have several buildings spread all over the city with the books distributed among them, you normally find this is when you start having troubles.